File: auth/src/main/java/com/sherpa/auth/security/userattributes/OAuthAttributes.java
package com.sherpa.auth.security.userattributes;

import java.util.Map;

import lombok.Builder;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@Builder
@RequiredArgsConstructor
public class OAuthAttributes {
	private final Map<String, Object> attributes;
	private final String name;
	private final String email;
	private final String profileURL;
}

File: gateway/src/main/kotlin/com/sherpa/gateway/domain/auth/oauthattributes/OAuthAttributes.kt
package com.sherpa.gateway.domain.auth.oauthattributes

import lombok.extern.slf4j.Slf4j
import org.slf4j.LoggerFactory

@Slf4j
class OAuthAttributes(
    val name: String,
    val email: String,
    val profileURL: String
) {

    companion object {
        private val logger = LoggerFactory.getLogger(OAuthAttributes::class.java)

        fun of(attributes: Map<String, Any>, provider: OAuthProvider): OAuthAttributes {

            try {
                return when (provider) {
                    OAuthProvider.GOOGLE -> ofGoogle(attributes)
                    OAuthProvider.GITHUB -> throw RuntimeException("not yet")
                }
            } catch (e: Exception) {
                logger.error("Error creating OAuthAttributes from attributes: $attributes", e)
                throw e
            }
        }

        private fun ofGoogle(attributes: Map<String, Any>): OAuthAttributes {
            return OAuthAttributes(
                name = attributes["name"] as String,
                email = attributes["email"] as String,
                profileURL = attributes["picture"] as String
            )
        }
    }


}

